staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/interface.go

```diff
@@ -0,0 +1,89 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fairqueuing
+
+import (
+	"context"
+	"time"
+)
+
+// QueueSetFactory is used to create QueueSet objects.
+type QueueSetFactory interface {
+	NewQueueSet(config QueueSetConfig) (QueueSet, error)
+}
+
+// QueueSet is the abstraction for the queuing and dispatching
+// functionality of one non-exempt priority level.  It covers the
+// functionality described in the "Assignment to a Queue", "Queuing",
+// and "Dispatching" sections of
+// https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md
+// .  Some day we may have connections between priority levels, but
+// today is not that day.
+type QueueSet interface {
+	// SetConfiguration updates the configuration
+	SetConfiguration(QueueSetConfig) error
+
+	// Quiesce controls whether the QueueSet is operating normally or is quiescing.
+	// A quiescing QueueSet drains as normal but does not admit any
+	// new requests. Passing a non-nil handler means the system should
+	// be quiescing, a nil handler means the system should operate
+	// normally. A call to Wait while the system is quiescing
+	// will be rebuffed by returning tryAnother=true. If all the
+	// queues have no requests waiting nor executing while the system
+	// is quiescing then the handler will eventually be called with no
+	// locks held (even if the system becomes non-quiescing between the
+	// triggering state and the required call).
+	Quiesce(EmptyHandler)
+
+	// Wait uses the given hashValue as the source of entropy
+	// as it shuffle-shards a request into a queue and waits for
+	// a decision on what to do with that request.  If tryAnother==true
+	// at return then the QueueSet has become undesirable and the client
+	// should try to find a different QueueSet to use; execute and
+	// afterExecution are irrelevant in this case.  Otherwise, if execute
+	// then the client should start executing the request and, once the
+	// request finishes execution or is canceled, call afterExecution().
+	// Otherwise the client should not execute the
+	// request and afterExecution is irrelevant.
+	Wait(ctx context.Context, hashValue uint64) (tryAnother, execute bool, afterExecution func())
+}
+
+// QueueSetConfig defines the configuration of a QueueSet.
+type QueueSetConfig struct {
+	// Name is used to identify a queue set, allowing for descriptive information about its intended use
+	Name string
+	// ConcurrencyLimit is the maximum number of requests of this QueueSet that may be executing at a time
+	ConcurrencyLimit int
+	// DesiredNumQueues is the number of queues that the API says should exist now; this may be smaller
```

Comment:
Sorry, I was not thinking straight when I suggested "this may be smaller...".  That was a remark for an implementation state variable, not this field of the struct that is a general definition of desired configuration.

Reply:
Removed the section after the semi-colon:
```
// DesiredNumQueues is the number of queues that the API says should exist now
```